### PR TITLE
protocol: add cursor_shape

### PR DESCRIPTION
**Pull request by kchibisov**

### CODE REVIEW

**Comment by kchibisov** [comment]
@YaLTeR should probably work now.

**Comment by YaLTeR** [comment]
Okay, I believe I got everything working now.

**Comment by YaLTeR** [comment]
Oh, looks like I accidentally erased the commit author. How do I get it back easily?

File: src/cursor.rs [hunk 1]
```diff
@@ -3,116 +3,310 @@ use std::collections::HashMap;
...
+use smithay::utils::{IsAlive, Logical, Physical, Point, Transform};
+use smithay::wayland::compositor::with_states;
 use xcursor::parser::{parse_xcursor, Image};
 use xcursor::CursorTheme;
 
+/// Some default looking `left_ptr` icon.
 static FALLBACK_CURSOR_DATA: &[u8] = include_bytes!("../resources/cursor.rgba");
 
-type CursorCache = HashMap<i32, (TextureBuffer<GlesTexture>, Point<i32, Physical>)>;
+type XCursorCache = HashMap<(CursorIcon, i32), Option<Rc<XCursor>>>;
```
[the remark this answers is not in the document]
Comment (YaLTeR) [reply src/cursor.rs:23]: I needed `&self` in `get()` signature, so I need this inside a `RefCell`. So the `get()` would need to return a `Ref` or `RefMut`. But I couldn't map `RefMut<HashMap>` into a `RefMut<Option<&XCursor>>`, only a `RefMut<Option<XCursor>>`, which is not what I wanted. So I changed it to an `Rc`. Basically, battle lost with the borrow checker.

File: src/niri.rs [hunk 2]
```diff
@@ -1414,6 +1433,11 @@ impl Niri {
             .unwrap()
             .are_animations_ongoing();
 
+        // Also keep redrawing if the current cursor is animated.
+        state.unfinished_animations_remain |= self
+            .cursor_manager
+            .is_current_cursor_animated(output.current_scale().integer_scale());
```
[the remark this answers is not in the document]
Comment (YaLTeR) [reply src/niri.rs:1439]: Pretty sure that code path always triggers even for outputs where the pointer currently is not located. Selective redrawing for pointer stuff is something I've been planning to tackle for a long time, but it's not done yet.

File: src/cursor.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/cursor.rs:23]: you see, `RefCell` is evil, could just have `&mut` and it'll be more or less fine...

**Review by kchibisov — COMMENTED** [review]
I guess it's fine then.

**Comment by YaLTeR** [comment]
Thanks

**Pull request merged**